### PR TITLE
I added the feature importance function

### PR DESCRIPTION
**Pull request by zelioluca**
Hello Niko! How are you? Let's start with this that is simple :-)

### CODE REVIEW

**Comment by zelioluca** [comment]
I passed the pre commit stuff already

**Comment by nmaarnio** [comment]
Hi, thanks for the updates! I'll review it now.

File: eis_toolkit/feature_importance/feature_importance.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+from sklearn.inspection import permutation_importance
+
+from eis_toolkit.exceptions import InvalidDatasetException
+
+
+def evaluate_feature_importance(
+    clf: sklearn.neural_network or sklearn.linear_model,
+    x_test: np.ndarray,
+    y_test: np.ndarray,
+    feature_names: list[str],
```
Comment (nmaarnio) [inline eis_toolkit/feature_importance/feature_importance.py:14]: Is any sequence ok here? We have used `Sequence[str]` for many cases before to allow tuples etc.

File: eis_toolkit/feature_importance/feature_importance.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    number_of_repetition: int = 50,
+    random_state: int = 0,
```
Comment (nmaarnio) [inline eis_toolkit/feature_importance/feature_importance.py:16]: Should the default be 0 or None? I thought None would be good default as it is so in Sklearn, but just asking

**Comment by zelioluca** [comment]
Let s see let s see 😁

File: eis_toolkit/feature_importance/feature_importance.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import numpy as np
+import pandas
```
Comment (nmaarnio) [inline eis_toolkit/feature_importance/feature_importance.py:2]: We have used `import pandas as pd` so far (and I noticed there seems to be duplicate import). Sorry for nitcpicky comment, but if its ok you should use the same import alias as elsewhere

File: eis_toolkit/feature_importance/feature_importance.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    y_test: np.ndarray,
+    feature_names: list[str],
+    number_of_repetition: int = 50,
+    random_state: int = 0,
+) -> (pandas.DataFrame, dict):
+    """
+    Evaluate the feature importance of a sklearn classifier or linear model.
+
+    Parameters:
+        clf (Any sklearn nn model or lm model): Trained classifier.
```
Comment (nmaarnio) [inline eis_toolkit/feature_importance/feature_importance.py:22]: Regarding the docstring, no need to include the type in paranthesis (just a decision we made).

File: eis_toolkit/feature_importance/feature_importance.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        x_test (np.ndarray): Testing feature data (X data need to be normalized / standardized).
+        y_test (np.ndarray): Testing target data.
+        feature_names (list): Names of the feature columns.
+        number_of_repetition (int): Number of iteration used when calculate feature importance (default 50).
+        random_state (int): random state for repeatability of results (Default 0).
+    Return:
```
Comment (nmaarnio) [inline eis_toolkit/feature_importance/feature_importance.py:28]:
```suggestion
        random_state (int): random state for repeatability of results (Default 0).
    
    Return:
```

File: eis_toolkit/feature_importance/feature_importance.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        feature_importance (pd.Dataframe): A dataframe composed by features name and Importance value
+        result (dict[object]): The resulted object with importance mean, importance std, and overall importance
+    Raise:
```
Comment (nmaarnio) [inline eis_toolkit/feature_importance/feature_importance.py:31]:
```suggestion
    
    Raises:
```

File: eis_toolkit/feature_importance/feature_importance.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+    result = permutation_importance(
+        clf, x_test, y_test.ravel(), n_repeats=number_of_repetition, random_state=random_state
+    )
+
+    feature_importance = pd.DataFrame({"Feature": feature_names, "Importance": result.importances_mean})
+
+    feature_importance["Importance"] = feature_importance["Importance"] * 100
+    feature_importance = feature_importance.sort_values(by="Importance", ascending=False)
+    # feature_importance['Importance'] = feature_importance['Importance'].apply(lambda x: '{:.6f}%'.format(x))
```
Comment (nmaarnio) [inline eis_toolkit/feature_importance/feature_importance.py:46]: Could remove this if its not needed for final implementation

File: eis_toolkit/feature_importance/feature_importance.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    Parameters:
+        clf (Any sklearn nn model or lm model): Trained classifier.
+        x_test (np.ndarray): Testing feature data (X data need to be normalized / standardized).
+        y_test (np.ndarray): Testing target data.
+        feature_names (list): Names of the feature columns.
+        number_of_repetition (int): Number of iteration used when calculate feature importance (default 50).
+        random_state (int): random state for repeatability of results (Default 0).
+    Return:
+        feature_importance (pd.Dataframe): A dataframe composed by features name and Importance value
+        result (dict[object]): The resulted object with importance mean, importance std, and overall importance
```
Comment (nmaarnio) [inline eis_toolkit/feature_importance/feature_importance.py:30]: The names of the returned objects don't need to be stated, you can check some other docstrings for examples.

File: tests/feature_importance/feature_importance.py
Comment (nmaarnio) [inline tests/feature_importance/feature_importance.py:1]: The test file should use `pytest` and have some basic unit type tests. These are sometimes quite stupid to do for models, but you could load some test data or create some small test array and then see that the results are as you would expect. Again I suggest to take a look at some other test files.

Some features are nice to demonstrate as a notebook, but if this does not deserve its own one in the toolkit (up to you), we can combine some modeling related stuff in one notebook to verify they work and show people how to use them at the same time.

File: .idea/eis_toolkit.iml
Comment (nmaarnio) [inline .idea/eis_toolkit.iml:1]: Not sure what is the `.idea` folder, but you should probably remove it from the PR

File: eis_toolkit/feature_importance/feature_importance.py
Comment (nmaarnio) [inline eis_toolkit/feature_importance/feature_importance.py:1]: Feature importance is marked in the excel at least to belong to Exploratory analyses, so you should move it there. Or if you think some other is more suitable thats ok too, but it shouldnt have its own folder directly under eis_toolkit folder

**Review by nmaarnio — CHANGES_REQUESTED** [review]
Hi! The core functionality looks good to me and overall the style is almost according to the guideline, but I found a few small nitpicky style issues still :) After you fix those and the test, should be good to go!

**Comment by zelioluca** [comment]
Hello there i do it now 😁

File: eis_toolkit/feature_importance/feature_importance.py (hunk 2 above)
Comment (zelioluca) [reply eis_toolkit/feature_importance/feature_importance.py:16]: this is an int should be any number I always set it to 0

File: eis_toolkit/feature_importance/feature_importance.py (hunk 4 above)
Comment (zelioluca) [reply eis_toolkit/feature_importance/feature_importance.py:22]: I removed type than!

**Comment by zelioluca** [comment]
Hi made all except pytest! Let me think little bit whatvi can do here 😁
anything else is complete

**Comment by nmaarnio** [comment]
Forgot one thing from the review: please add the `@beartype` decorator for input checking!

File: eis_toolkit/feature_importance/__init__.py
Comment (nmaarnio) [inline eis_toolkit/feature_importance/__init__.py:1]: Also please delete this folder

File: tests/exploratory_analyses/feature_importance_test.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    "Rd_TC",
+    "Rd_Th",
+    "Rd_K",
+    "EM_ratio",
+    "EM_Ap_rs",
+    "Em_Qd",
+    "EM_Inph",
+]
+
+X = pd.read_csv("../data/local/data/fake_smote_data.csv").to_numpy()
```
Comment (nmaarnio) [inline tests/exploratory_analyses/feature_importance_test.py:31]: If you put this dataset to somewhere in the data/remote, then the test will work for everyone. Now the dataset stays on your computer.

**Review by nmaarnio — COMMENTED** [review]
Only a few files/folder should be removed from the PR and test data put to remote, then we can merge!

**Comment by zelioluca** [comment]
Give me file

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

File: tests/exploratory_analyses/feature_importance_test.py (hunk 9 above)
Comment (zelioluca) [reply tests/exploratory_analyses/feature_importance_test.py]: I puty it in remote folder

**Comment by zelioluca** [comment]
I m doing the test now

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by zelioluca** [comment]
How i put the path of that folder like this test does not like

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by zelioluca** [comment]
Yes i fixed it 😁

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by zelioluca** [comment]
tests passed finally

**Comment by nmaarnio** [comment]
Looks good now, merging!

**Review by nmaarnio — APPROVED** [review]

**Comment by zelioluca** [comment]
Tomorrow i do cnn

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Pull request merged**